### PR TITLE
Fix: Also replace tags in the list of tags referenced from outside.

### PR DESCRIPTION
**Pull request by chriseth**
Hopefully the last fix needed by #5102

### CODE REVIEW

File: libevmasm/Assembly.cpp [hunk 1]
```diff
@@ -436,7 +436,17 @@ map<u256, u256> Assembly::optimiseInternal(
 			BlockDeduplicator dedup{m_items};
 			if (dedup.deduplicate())
 			{
-				tagReplacements.insert(dedup.replacedTags().begin(), dedup.replacedTags().end());
+				for (auto const& replacement: dedup.replacedTags())
+				{
+					tagReplacements[replacement.first] = replacement.second;
+					assertThrow(
```
Comment (axic) [inline libevmasm/Assembly.cpp:441]: Swap these two statements?

File: libevmasm/Assembly.cpp [hunk 2]
```diff
@@ -436,7 +436,17 @@ map<u256, u256> Assembly::optimiseInternal(
 			BlockDeduplicator dedup{m_items};
 			if (dedup.deduplicate())
 			{
-				tagReplacements.insert(dedup.replacedTags().begin(), dedup.replacedTags().end());
+				for (auto const& replacement: dedup.replacedTags())
+				{
+					tagReplacements[replacement.first] = replacement.second;
```
Comment (axic) [inline libevmasm/Assembly.cpp]: Should this assert that `replacement.first` doesn't exist yet?

File: Changelog.md [hunk 3]
```diff
@@ -13,6 +13,7 @@ Bugfixes:
  * ABIEncoderV2: Fix internal error related to ecrecover.
  * ABIEncoderV2: Fix internal error related to mappings as library parameters.
  * Inline Assembly: Proper error message for missing variables.
+ * Optimizer: Fix bug related to unused tag removal across assemblies.
```
Comment (axic) [inline Changelog.md]: If this never had any adverse effect we should mention that.

File: libevmasm/Assembly.cpp (hunk 2 above)
Comment (chriseth) [reply libevmasm/Assembly.cpp]: We can try and see what happens...

File: Changelog.md (hunk 3 above)
Comment (chriseth) [reply Changelog.md]: Reworded into mentioning "internal error", I think this should be sufficient. It always resulted in a "tag not found" - tag numbers are not re-used, only replaced.

Comment (axic) [reply Changelog.md]: I think we should still mention it didn't generate invalid code to avoid scaring people.

Comment (chriseth) [reply Changelog.md]: Added.

File: libevmasm/Assembly.cpp [hunk 4]
```diff
@@ -390,7 +390,7 @@ Assembly& Assembly::optimise(OptimiserSettings const& _settings)
 
 map<u256, u256> Assembly::optimiseInternal(
 	OptimiserSettings const& _settings,
-	std::set<size_t> const& _tagsReferencedFromOutside
+	std::set<size_t> _tagsReferencedFromOutside
```
Comment (leonardoalt) [inline libevmasm/Assembly.cpp:393]: Shouldn't it be `&`?

Comment (chriseth) [reply libevmasm/Assembly.cpp:393]: No, it is not needed in the caller. The function returns the replacements themselves and then the caller takes them into account.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**